[UI-side compositing] Legacy scrollbars do not flash grey when hovered over

nmoucht opened the pull request:
#### 15cb0cae9dcf910ce6862066f42cb29fa9e09741
<pre>
[UI-side compositing] Legacy scrollbars do not flash grey when hovered over
<a href="https://bugs.webkit.org/show_bug.cgi?id=254721">https://bugs.webkit.org/show_bug.cgi?id=254721</a>
rdar://107447969

Reviewed by Simon Fraser.

For legacy scrollbars we need to call NSScrollerImp functions to notify them when a mouse
is hovering over them. The approach in this patch is to plumb a pair of bits representing
if a mouse is in the vertical or horizontal scrollbar through the scrolling tree, which
eventually calls into the scroller imp of ScrollerMac. Since this is only needed for legacy
scrollbars return early in ScrollerMac::mouseEntered/ExitedScrollbar if the style is not
legacy.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateLastScrollbarUnderMouse):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::mouseDidEnterScrollbar):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::mouseDidEnterScrollbar):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
(WebCore::ScrollingStateScrollingNode::setMouseDidEnterScrollbar):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollingStateScrollingNode::mouseDidEnterScrollbar const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(WebCore::ScrollerPairMac::mouseEnteredScrollbar):
(WebCore::ScrollerPairMac::mouseExitedScrollbar):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebCore/platform/ScrollbarsController.h:
* Source/WebCore/platform/mac/NSScrollerImpDetails.mm:
(WebCore::ScrollerStyle::recommendedScrollerStyle):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder&lt;ScrollingStateScrollingNode&gt;::encode):
(ArgumentCoder&lt;ScrollingStateScrollingNode&gt;::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: <a href="https://commits.webkit.org/262721@main">https://commits.webkit.org/262721@main</a>
</pre>
<!--EWS-Status-Bubble-Start-->
https://github.com/WebKit/WebKit/commit/4fdb5052f5162bb12713b5a8c70c6988cd84a5f2

| Misc | iOS, tvOS & watchOS  | macOS  | Linux |  Windows |
| ----- | ---------------------- | ------- |  ----- |  --------- |
| [❌ 🧪 style](https://ews-build.webkit.org/#/builders/38/builds/2315 "1 style error") | [✅ 🛠 ios](https://ews-build.webkit.org/#/builders/26/builds/2327 "Built successfully") | [✅ 🛠 mac](https://ews-build.webkit.org/#/builders/14/builds/2428 "Built successfully") | [✅ 🛠 wpe](https://ews-build.webkit.org/#/builders/5/builds/3260 "Built successfully") | [✅ 🛠 wincairo](https://ews-build.webkit.org/#/builders/32/builds/2343 "Built successfully") 
| [✅ 🧪 bindings](https://ews-build.webkit.org/#/builders/9/builds/2288 "Passed tests") | [✅ 🛠 ios-sim](https://ews-build.webkit.org/#/builders/23/builds/2421 "Built successfully") | [✅ 🛠 mac-AS-debug](https://ews-build.webkit.org/#/builders/16/builds/2402 "Built successfully") | [✅ 🧪 wpe-wk2](https://ews-build.webkit.org/#/builders/34/builds/2084 "Passed tests") | 
| [✅ 🧪 webkitperl](https://ews-build.webkit.org/#/builders/11/builds/2337 "Passed tests") | [✅ 🧪 ios-wk2](https://ews-build.webkit.org/#/builders/15/builds/2044 "Passed tests") | [✅ 🧪 api-mac](https://ews-build.webkit.org/#/builders/18/builds/2085 "Passed tests") | [✅ 🛠 gtk](https://ews-build.webkit.org/#/builders/2/builds/3140 "Built successfully") | 
| | [✅ 🧪 ios-wk2-wpt](https://ews-build.webkit.org/#/builders/39/builds/31 "Passed tests") | [✅ 🧪 mac-wk1](https://ews-build.webkit.org/#/builders/10/builds/2070 "Passed tests") | [✅ 🧪 gtk-wk2](https://ews-build.webkit.org/#/builders/1/builds/1946 "Passed tests") | 
| | [✅ 🛠 tv](https://ews-build.webkit.org/#/builders/7/builds/2146 "Built successfully") | [✅ 🧪 api-ios](https://ews-build.webkit.org/#/builders/13/builds/2071 "Passed tests") | [✅ 🧪 api-gtk](https://ews-build.webkit.org/#/builders/21/builds/3259 "Passed tests") | 
| | [✅ 🧪 mac-wk2](https://ews-build.webkit.org/#/builders/36/builds/2118 "Passed tests") | [✅ 🧪 mac-AS-debug-wk2](https://ews-build.webkit.org/#/builders/17/builds/1908 "Passed tests") | | 
| | [✅ 🛠 tv-sim](https://ews-build.webkit.org/#/builders/30/builds/2123 "Built successfully") | [✅ 🧪 mac-wk2-stress](https://ews-build.webkit.org/#/builders/8/builds/2076 "Passed tests") | | 
| | [✅ 🛠 watch](https://ews-build.webkit.org/#/builders/4/builds/2054 "Built successfully") | | | 
| [✅ 🛠 🧪 unsafe-merge](https://ews-build.webkit.org/#/builders/22/builds/271 "Built successfully and passed tests") | [✅ 🛠 watch-sim](https://ews-build.webkit.org/#/builders/31/builds/2245 "Built successfully") | | | 
<!--EWS-Status-Bubble-End-->

shivamidow commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4673,11 +4673,19 @@ void EventHandler::updateLastScrollbarUnderMouse(Scrollbar* scrollbar, SetOrClea
         // Send mouse exited to the old scrollbar.
-        if (m_lastScrollbarUnderMouse)
+        
+        if (m_lastScrollbarUnderMouse){
```
/* Unofficial Review */
Missing space after ')'

nmoucht replied:
Fixed.

shivamidow commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -302,6 +302,24 @@ bool AsyncScrollingCoordinator::requestStopKeyboardScrollAnimation(ScrollableAre
     return true;
 }
 
+void AsyncScrollingCoordinator::mouseDidEnterScrollbar(ScrollableArea& scrollableArea, bool didEnterScrollbar)
```
/* Unofficial Review */
I think either removing didEnterScrollbar or changing the method name is needed. The method name says the mouse entered scrollbar though, why do we need a boolean for it?

nmoucht replied:
I think this name is reasonable as the bool represents if the mouse is entering or exiting the scrollbar.

shivamidow commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -275,6 +276,17 @@ void ScrollingStateScrollingNode::setScrollerImpsFromScrollbars(Scrollbar*, Scro
+    if (flag == m_mouseDidEnterScrollbar)
+        return;
+
+#if PLATFORM(MAC)
```
/* Unofficial Review */
Do we need this if guard? Other ports do not handle Property::MouseDidEnterScrollbar.

nmoucht replied:
Removed.

shivamidow commented on Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
```diff
@@ -68,6 +68,10 @@
     }
     
+    if (scrollingStateNode.hasChangedProperty(ScrollingStateNode::Property::MouseDidEnterScrollbar)) {
+        scrollingStateNode.mouseDidEnterScrollbar() ? m_scrollerPair->mouseEnteredScrollbar() : m_scrollerPair->mouseExitedScrollbar();
```
/* Unofficial Review */
No return value is taken. Please consider if else instead.

nmoucht replied:
This part changed under the new patch.

shivamidow commented on Source/WebCore/platform/ScrollbarsController.h:
```diff
@@ -29,6 +29,7 @@
 #include "FloatSize.h"
 #include <wtf/FastMalloc.h>
 #include <wtf/text/WTFString.h>
+#include "ScrollableArea.h"
```
Maybe, a clean-up mistake?

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/page/EventHandler.cpp:
```diff
@@ -4673,11 +4673,16 @@ void EventHandler::updateLastScrollbarUnderMouse(Scrollbar* scrollbar, SetOrClea
-        if (m_lastScrollbarUnderMouse)
+        if (m_lastScrollbarUnderMouse) {
+            if (auto scrollingCoordinator = m_frame.page()->scrollingCoordinator())
+                scrollingCoordinator->mouseIsInScrollbar(m_lastScrollbarUnderMouse->scrollableArea(), false, m_lastScrollbarUnderMouse->orientation());
```
shouldn't this be called `setMouseIsInScrollbar()`?

cdumez replied:
Have you considered passing the scrollbar as parameter instead of both its scrollableArea and its orientation?

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -302,6 +302,23 @@ bool AsyncScrollingCoordinator::requestStopKeyboardScrollAnimation(ScrollableAre
+
+    stateNode->setMouseIsInScrollbar({ orientation == ScrollbarOrientation::Vertical ? didEnterScrollbar : false, orientation == ScrollbarOrientation::Vertical ? false : didEnterScrollbar });
+    scrollingStateTreePropertiesChanged();
+    return;
```
This return is not needed.

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -275,6 +276,15 @@ void ScrollingStateScrollingNode::setScrollerImpsFromScrollbars(Scrollbar*, Scro
 }
 #endif
 
+void ScrollingStateScrollingNode::setMouseIsInScrollbar(std::pair<bool, bool> flags)
```
This std::pair may read better as a struct. It is very unclear that it represents right now.

nmoucht replied:
Changed to struct.

cdumez commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -381,6 +381,27 @@ - (void)invalidate
+
+    [m_scrollerImp mouseEnteredScroller];
+}
+void ScrollerMac::mouseExitedScrollbar()
```
Missing blank line before this.

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
```diff
@@ -101,17 +101,23 @@ class ScrollerPairMac : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<S
 
     NSScrollerImp *scrollerImpHorizontal() { return horizontalScroller().scrollerImp(); }
     NSScrollerImp *scrollerImpVertical() { return verticalScroller().scrollerImp(); }
+    
```
Unnecessary extra line.

nmoucht replied:
Fixed.

cdumez commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
```diff
@@ -101,17 +101,23 @@ class ScrollerPairMac : public ThreadSafeRefCountedAndCanMakeThreadSafeWeakPtr<S
     void setVerticalScrollerImp(NSScrollerImp *);
     void setHorizontalScrollerImp(NSScrollerImp *);
+    
+    NSScrollerImpPair* scrollerImpPair() const { return m_scrollerImpPair.get(); }
```
star on wrong side.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -302,6 +302,22 @@ bool AsyncScrollingCoordinator::requestStopKeyboardScrollAnimation(ScrollableAre
+        return;
+
+    stateNode->setMouseIsInScrollbar({ scrollbar->orientation() == ScrollbarOrientation::Vertical ? didEnterScrollbar : false, scrollbar->orientation() == ScrollbarOrientation::Vertical ? false : didEnterScrollbar });
+    scrollingStateTreePropertiesChanged();
```
This shouldn’t be necessary. `ScrollingStateTree::setHasChangedProperties()` calls this.

nmoucht replied:
Removed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
```diff
@@ -167,6 +167,8 @@ class AsyncScrollingCoordinator : public ScrollingCoordinator {
     void wheelEventScrollWillStartForNode(ScrollingNodeID);
     void wheelEventScrollDidEndForNode(ScrollingNodeID);
+    
+    WEBCORE_EXPORT void setMouseIsInScrollbar(Scrollbar*, bool didEnterScrollbar) override;
```
`didEnterScrollbar` -> `isOverScrollbar`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateNode.h:
```diff
@@ -225,37 +225,38 @@ enum class ScrollingStateNodeProperty : uint64_t {
     HorizontalScrollbarLayer                    = 1LLU << 16,
     VerticalScrollbarLayer                      = 1LLU << 17,
     PainterForScrollbar                         = 1LLU << 18,
+    MouseDidEnterScrollbar                      = 1LLU << 19,
```
MouseIsScrollbarHoverState? Or just `ScrollbarHoverState`

nmoucht replied:
Changed to ScrollbarHoverState.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -48,6 +48,7 @@ ScrollingStateScrollingNode::ScrollingStateScrollingNode(const ScrollingStateScr
     , m_scrollOrigin(stateNode.scrollOrigin())
     , m_snapOffsetsInfo(stateNode.m_snapOffsetsInfo)
 #if PLATFORM(MAC)
+    , m_mouseIsInScrollbar(stateNode.mouseIsInScrollbar())
```
`m_scrollbarHoverState`?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
```diff
@@ -38,6 +38,12 @@ OBJC_CLASS NSScrollerImp;
 
 namespace WebCore {
 
+struct MouseIsInScrollbar {
```
`ScrollbarHoverState`. The current name makes it sound like it only applies to one scrollbar.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
```diff
@@ -38,6 +38,12 @@ OBJC_CLASS NSScrollerImp;
 namespace WebCore {
 
+struct MouseIsInScrollbar {
+    bool mouseIsInVerticalScrollbar;
```
Initialize these to false.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -275,6 +276,15 @@ void ScrollingStateScrollingNode::setScrollerImpsFromScrollbars(Scrollbar*, Scro
 }
 #endif
 
+void ScrollingStateScrollingNode::setMouseIsInScrollbar(MouseIsInScrollbar flags)
```
`setScrollbarHoverState`

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
```diff
@@ -164,6 +164,12 @@ void ArgumentCoder<ScrollingStateScrollingNode>::encode(Encoder& encoder, const
+    
+    if (node.hasChangedProperty(ScrollingStateNode::Property::MouseDidEnterScrollbar)) {
+        auto mouseIsInScrollbar = node.mouseIsInScrollbar();
+        encoder << mouseIsInScrollbar.mouseIsInVerticalScrollbar;
```
Do horizontal first.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
```diff
@@ -285,6 +292,16 @@ bool ArgumentCoder<ScrollingStateScrollingNode>::decode(Decoder& decoder, Scroll
+    
+    if (node.hasChangedProperty(ScrollingStateNode::Property::MouseDidEnterScrollbar)) {
+        bool didEnterScrollbarVertical;
+        if (!decoder.decode(didEnterScrollbarVertical))
```
Horizonal first.

nmoucht replied:
Fixed.

smfr commented on Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:
```diff
@@ -3011,6 +3011,7 @@ header: <WebCore/ScrollingStateNode.h>
     HorizontalScrollbarLayer
     VerticalScrollbarLayer
     PainterForScrollbar
+    MouseDidEnterScrollbar
```
Naming

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
```diff
@@ -302,6 +302,20 @@ bool AsyncScrollingCoordinator::requestStopKeyboardScrollAnimation(ScrollableAre
     return true;
 }
 
+void AsyncScrollingCoordinator::setMouseIsOverScrollbar(Scrollbar* scrollbar, bool didEnterScrollbar)
```
`didEnterScrollbar` -> `isOverScrollbar`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
```diff
@@ -285,6 +286,15 @@ void ScrollingStateScrollingNode::setMouseIsOverContentArea(bool flag)
     setPropertyChanged(Property::ContentAreaHoverState);
 }
 
+void ScrollingStateScrollingNode::setScrollbarHoverState(ScrollbarHoverState flags)
```
`flags` -> `hoverState`

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
```diff
@@ -38,6 +38,13 @@ OBJC_CLASS NSScrollerImp;
 namespace WebCore {
 
+struct ScrollbarHoverState {
+    bool mouseIsInHorizontalScrollbar { false };
```
Maybe “is over” rather than “is in"

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
```diff
@@ -132,6 +142,8 @@ class ScrollingStateScrollingNode : public ScrollingStateNode {
     LayerRepresentation m_horizontalScrollbarLayer;
     LayerRepresentation m_verticalScrollbarLayer;
+    
+    ScrollbarHoverState m_scrollbarHoverState { false, false };
```
No need for the explicit initialization.

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerMac.mm:
```diff
@@ -381,6 +381,30 @@ - (void)invalidate
+    if ([m_pair.scrollerImpPair() overlayScrollerStateIsLocked])
+        return;
+
+    [m_scrollerImp mouseEnteredScroller];
```
Does this need to bounce to main thread?

nmoucht replied:
Fixed.

smfr commented on Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
```diff
@@ -366,6 +366,23 @@ - (void)scrollerImpPair:(NSScrollerImpPair *)scrollerImpPair updateScrollerStyle
     [m_scrollerImpPair mouseExitedContentArea];
 }
 
+void ScrollerPairMac::mouseIsInScrollbar(ScrollbarHoverState flags)
```
`flags` -> `hoverState`

nmoucht replied:
Fixed.

nmoucht committed:
[UI-side compositing] Legacy scrollbars do not flash grey when hovered over
https://bugs.webkit.org/show_bug.cgi?id=254721
rdar://107447969

Reviewed by Simon Fraser.

For legacy scrollbars we need to call NSScrollerImp functions to notify them when a mouse
is hovering over them. The approach in this patch is to plumb a pair of bits representing
if a mouse is in the vertical or horizontal scrollbar through the scrolling tree, which
eventually calls into the scroller imp of ScrollerMac. Since this is only needed for legacy
scrollbars return early in ScrollerMac::mouseEntered/ExitedScrollbar if the style is not
legacy.

* Source/WebCore/page/EventHandler.cpp:
(WebCore::EventHandler::updateLastScrollbarUnderMouse):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.cpp:
(WebCore::AsyncScrollingCoordinator::mouseDidEnterScrollbar):
* Source/WebCore/page/scrolling/AsyncScrollingCoordinator.h:
* Source/WebCore/page/scrolling/ScrollingCoordinator.h:
(WebCore::ScrollingCoordinator::mouseDidEnterScrollbar):
* Source/WebCore/page/scrolling/ScrollingStateNode.h:
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.cpp:
(WebCore::ScrollingStateScrollingNode::ScrollingStateScrollingNode):
(WebCore::ScrollingStateScrollingNode::setMouseDidEnterScrollbar):
* Source/WebCore/page/scrolling/ScrollingStateScrollingNode.h:
(WebCore::ScrollingStateScrollingNode::mouseDidEnterScrollbar const):
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.h:
* Source/WebCore/page/scrolling/mac/ScrollerPairMac.mm:
(WebCore::ScrollerPairMac::mouseEnteredScrollbar):
(WebCore::ScrollerPairMac::mouseExitedScrollbar):
* Source/WebCore/page/scrolling/mac/ScrollingTreeScrollingNodeDelegateMac.mm:
(WebCore::ScrollingTreeScrollingNodeDelegateMac::updateFromStateNode):
* Source/WebCore/platform/ScrollbarsController.h:
* Source/WebCore/platform/mac/NSScrollerImpDetails.mm:
(WebCore::ScrollerStyle::recommendedScrollerStyle):
* Source/WebKit/Shared/RemoteLayerTree/RemoteScrollingCoordinatorTransaction.cpp:
(ArgumentCoder<ScrollingStateScrollingNode>::encode):
(ArgumentCoder<ScrollingStateScrollingNode>::decode):
* Source/WebKit/Shared/WebCoreArgumentCoders.serialization.in:

Canonical link: https://commits.webkit.org/262721@main
